File: security-platform/agents/ios/Sources/SecurityPlatform/PolicyEngine.swift
import Foundation

/// Policy engine for local rule evaluation.
public class PolicyEngine {
    private let config: AgentConfig.PolicyConfig

    public init(config: AgentConfig.PolicyConfig) {
        self.config = config
    }

    public func check(action: String, context: [String: Any]) -> PolicyResult {
        // In observe mode, always allow but log
        if config.mode == "observe" {
            return PolicyResult(allowed: true, reason: "observe_mode")
        }

        // In block mode, evaluate rules
        if config.mode == "block" {
            for rule in config.rules {
                if rule.action == action {
                    return evaluateRule(rule, context: context)
                }
            }
            return PolicyResult(allowed: true, reason: "no_matching_rule")
        }

        return PolicyResult(allowed: true, reason: "default_allow")
    }

    private func evaluateRule(_ rule: PolicyRule, context: [String: Any]) -> PolicyResult {
        if rule.effect == "deny" {
            return PolicyResult(allowed: false, reason: "rule_denied: \(rule.id)")
        }
        return PolicyResult(allowed: true, reason: "rule_allowed: \(rule.id)")
    }
}

/// Policy rule definition.
public struct PolicyRule {
    public var id: String
    public var action: String
    public var condition: String
    public var effect: String // "allow" or "deny"

    public init(id: String, action: String, condition: String, effect: String) {
        self.id = id
        self.action = action
        self.condition = condition
        self.effect = effect
    }
}

/// Result of a policy check.
public struct PolicyResult {
    public let allowed: Bool
    public let reason: String

    public init(allowed: Bool, reason: String) {
        self.allowed = allowed
        self.reason = reason
    }
}

File: security-platform/agents/java/src/main/java/com/securityplatform/agent/PolicyEngine.java
package com.securityplatform.agent;

import java.util.Map;

/**
 * Policy Engine for local policy evaluation
 */
public class PolicyEngine {
    private AgentConfig config;

    public PolicyEngine(AgentConfig config) {
        this.config = config;
    }

    public PolicyResult checkPolicy(String policyName, Map<String, Object> context) {
        // In observe mode, always allow but record
        if ("observe".equals(config.getLocalPolicy()) || config.getLocalPolicy() == null) {
            return new PolicyResult(true, "observe_mode");
        }

        // Simple policy checks (can be extended)
        // For now, always allow in block mode unless explicitly configured otherwise
        return new PolicyResult(true, "allowed");
    }

    public static class PolicyResult {
        private boolean allowed;
        private String reason;

        public PolicyResult(boolean allowed, String reason) {
            this.allowed = allowed;
            this.reason = reason;
        }

        public boolean isAllowed() {
            return allowed;
        }

        public String getReason() {
            return reason;
        }
    }
}
